src/components/views/messages/TextualBody.js

```diff
@@ -37,28 +38,76 @@ module.exports = React.createClass({
 
         /* link URL for the highlights */
         highlightLink: React.PropTypes.string,
+
+        /* callback for when our widget has loaded */
+        onWidgetLoad: React.PropTypes.func,
+    },
+
+    getInitialState: function() {
+        return {
+            link: null,
+
+            // track whether the preview widget is hidden
+            // we can't directly use mxEvent's widgetHidden property
+            // as shouldComponentUpdate needs to be able to do before & after
+            // comparisons of the property (and we don't pass it in as a top
+            // level prop to avoid bloating the number of props flying around)
+            widgetHidden: false,
+        };
     },
 
     componentDidMount: function() {
         linkifyElement(this.refs.content, linkifyMatrix.options);
 
-        if (this.props.mxEvent.getContent().format === "org.matrix.custom.html")
-            HtmlUtils.highlightDom(ReactDOM.findDOMNode(this));
-    },
+        var link = this.findLink(this.refs.content.children);
+        if (link) {
+            this.setState({ link: link.getAttribute("href") });
 
-    componentDidUpdate: function() {
-        // XXX: why don't we linkify here?
-        // XXX: why do we bother doing this on update at all, given events are immutable?
+            // lazy-load the hidden state of the preview widget from localstorage
+            if (global.localStorage) {
+                var hidden = global.localStorage.getItem("hide_preview_" + this.props.mxEvent.getId());
+                this.props.mxEvent.widgetHidden = hidden;
+                this.setState({ widgetHidden: hidden });
+            }
+        }
 
         if (this.props.mxEvent.getContent().format === "org.matrix.custom.html")
             HtmlUtils.highlightDom(ReactDOM.findDOMNode(this));
     },
 
-    shouldComponentUpdate: function(nextProps) {
+    findLink: function(nodes) {
+        for (var i = 0; i < nodes.length; i++) {
+            var node = nodes[i];
+            if (node.tagName === "A" && node.getAttribute("href")) {
+                return node;
+            }
+            else if (node.children && node.children.length) {
+                return this.findLink(node.children)
+            }
+        }
+    },
+
+    shouldComponentUpdate: function(nextProps, nextState) {
         // exploit that events are immutable :)
         return (nextProps.mxEvent.getId() !== this.props.mxEvent.getId() ||
                 nextProps.highlights !== this.props.highlights ||
-                nextProps.highlightLink !== this.props.highlightLink);
+                nextProps.highlightLink !== this.props.highlightLink ||
+                nextState.link !== this.state.link ||
+                nextProps.mxEvent.widgetHidden !== this.state.widgetHidden);
+    },
+
+    componentWillUpdate: function(nextProps, nextState) {
+        this.setState({ widgetHidden: nextProps.mxEvent.widgetHidden });
```

Comment:
yes, because MessageContextualMenu can't get at TextualBody, but it can get at mxEvent (when unhiding the wretched things)